Monitor for external changes to script file

jclarkeSTFC opened the pull request:
**Description of work**
When another application modifies a script that's open in Workbench, then the user will now get a notification. They can either reload the file, or keep their version open but unsaved. This is the same behaviour as Notepad++. To do this there are multiple components, monitoring the files (but not when it's Workbench doing the saving), reloading the file, and marking the file as
modified.

**Purpose of work**
<!-- Why has this work been done? If there is no linked issue please provide appropriate context for this work.
-->
Fixes #34981 

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

**Summary of work**
<!-- Please provide a short, high level description of the work that was done.
-->
- Monitor scripts for changes with a QFileSystemWatcher
- When saving the scripts from Workbench, temporarily disable the file watcher
- On focus (required some signal plumbing), prompt the user if the script has changed externally. They can either reload the file (new method), or they can mark the file as modified (new method).
- Cppcheck suppression added for the new signal on ScriptEditor. It looks like trying to emit a signal from C++ will give you a shadow function error in Cppcheck, at least in the version we're currently using.


**To test:**
Open some scripts in Workbench and another editor. Make changes in either Workbench or the other editor and see how Workbench handles the situation. There are two options when the script has changed externally, either reload the script or keep the Workbench version but marked as modified, would be good to test both those options are working correctly. Can also try deleting or renaming a script file that's open in Workbench.

<!-- Instructions for testing.
There should be sufficient instructions for someone unfamiliar with the application to test - unless a specific
reviewer is requested.
If instructions for replicating the fault are contained in the linked issue then it is OK to refer back to these.
-->

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

#### Gatekeeper ####

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

jclarkeSTFC committed:
Watch when script files are changed.

Use a QFileSystemWatcher to let us know when any of the script files
are changed (either by us or another application). To be decided when
we act on this information.

jclarkeSTFC committed:
Monitor for external changes to script file

When another application modifies a script that's open in Mantid, then
the user will now get a notification. They can either reload the file, or
keep their version open but unsaved. This is the same behaviour as Notepad++.
To do this there are multiple components, monitoring the files (but not when
it's Mantid doing the saving), reloading the file, and marking the file as
modified.

jclarkeSTFC committed:
Added release notes

jclarkeSTFC committed:
Rename editor focus signal

This is due to a cppcheck shadow variable error. Also added an entry to
_common.sip which was missing from earlier commits, I think these two
things are related, because it did work before without the changes to
the .sip file

jclarkeSTFC committed:
Handle case of file being deleted.

jclarkeSTFC committed:
Add suppression for new signal on ScriptEditor

It looks like a lot of the time cppcheck gives a shadow function error
when you emit a signal from a qt widget. This started happening when
cppcheck was upgraded recently.

jhaigh0 commented on qt/python/mantidqt/mantidqt/widgets/codeeditor/multifileinterpreter.py:
```diff
@@ -66,6 +66,12 @@ def __init__(self, font=None, default_content=None, parent=None, completion_enab
+        # Object to monitor file changes
+        self.file_watcher = QFileSystemWatcher()
+        self.file_watcher.fileChanged.connect(self.file_changed_event)
+        self.files_changed_unhandled = set()
+        self.monitor_files_changing = True
```
I think a comment above these two variables would help to make what's happening clearer

jclarkeSTFC replied:
I'll add a comment

jclarkeSTFC replied:
```suggestion
        # When monitor_files_changing is True, then any files that are detected
        # as modified by the QFileSystemWatcher will be added to the set. We 
        # need to disable the file monitoring when we want to save the file so 
        # we don't get notified about our own save.
        self.files_changed_unhandled = set()
        self.monitor_files_changing = True
```

jhaigh0 commented on qt/python/mantidqt/mantidqt/widgets/codeeditor/multifileinterpreter.py:
```diff
@@ -403,3 +443,32 @@ def zoom_out_all_tabs(self):
     def on_completion_change(self):
         for idx in range(self.editor_count):
             self.editor_at(idx).setCompletion(self.completion_enabled)
+
+    def focus_in(self, file):
+        self.check_file_for_external_modifications(file)
+
```
do we need this function if it just calls another with the same arguments?

jclarkeSTFC replied:
I could inline the function, but I thought it would be more descriptive to use a separate function, plus maybe someone will want to do something else inside focus_in(). I don't mind, could inline it if it's clearer.

jhaigh0 reviewed (commented):
Overall very nicely thought out, just had a few questions about small things

jclarkeSTFC commented on qt/python/mantidqt/mantidqt/widgets/codeeditor/interpreter.py:
```diff
@@ -149,6 +150,7 @@ def __init__(self, font=None, content=None, filename=None, parent=None, completi
 
         self.editor.modificationChanged.connect(self.sig_editor_modified)
         self.editor.fileNameChanged.connect(self.sig_filename_modified)
+        self.editor.editorFocusIn.connect(self.sig_focus_in)
```
I'm not sure of the reasoning exactly, but I followed the pattern of the existing signals, e.g. fileNameChanged.

jhaigh0 replied:
ah okay, that makes some sense

jclarkeSTFC commented on qt/python/mantidqt/mantidqt/widgets/codeeditor/multifileinterpreter.py:
```diff
@@ -96,6 +102,10 @@ def load_settings_from_config(self, config):
         self.completion_enabled = config.get("Editors", "completion_enabled", type=bool)
         self.on_completion_change()
 
+    def file_changed_event(self, filename):
+        if self.monitor_files_changing and filename not in self.files_changed_unhandled:
+            self.files_changed_unhandled.add(filename)
```
```suggestion
    def file_changed_event(self, filename):
        if self.monitor_files_changing:
            self.files_changed_unhandled.add(filename)
```

jclarkeSTFC commented on qt/python/mantidqt/mantidqt/widgets/codeeditor/multifileinterpreter.py:
```diff
@@ -347,12 +379,20 @@ def restore_session_tabs(self):
     def save_current_file(self):
         """Save the current file"""
-        self.current_editor().save(force_save=True)
+        self.monitor_files_changing = False
+        try:
+            self.current_editor().save(force_save=True)
+        finally:
+            self.monitor_files_changing = True
```
Yes, I wanted to make sure that if save() goes wrong then the flag gets left in the correct state, otherwise the file monitoring will be broken until Workbench is restarted.

jclarkeSTFC commented on qt/python/mantidqt/mantidqt/widgets/codeeditor/multifileinterpreter.py:
```diff
@@ -403,3 +443,32 @@ def zoom_out_all_tabs(self):
+        if file not in self.files_changed_unhandled:
+            return
+
+        self.files_changed_unhandled.remove(file)
+        if osp.isfile(file):
+            reload_button = QMessageBox.question(
```
At the top of the file there's a `os.path as osp`. Originally I wrote `os.path`, but then noticed the `osp`. I'd prefer to use `os.path` everywhere in the file, but I didn't want to overrule whoever likes `osp`.

jhaigh0 committed:
Update qt/python/mantidqt/mantidqt/widgets/codeeditor/multifileinterpreter.py

Co-authored-by: James Clarke <139879523+jclarkeSTFC@users.noreply.github.com>

jhaigh0 committed:
Update qt/python/mantidqt/mantidqt/widgets/codeeditor/multifileinterpreter.py

Co-authored-by: James Clarke <139879523+jclarkeSTFC@users.noreply.github.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jhaigh0 commented on qt/python/mantidqt/mantidqt/widgets/codeeditor/multifileinterpreter.py:
```diff
@@ -403,3 +447,32 @@ def zoom_out_all_tabs(self):
+            else:
+                self.current_editor().mark_as_modified()
+        else:
+            QMessageBox.warning(
+                self,
+                "",
+                f"The current file ({file}) doesn't exist anymore, which is due to an external source deleting or renaming the \
+                    file. The contents of this tab will remain unchanged, but are unsaved.",
+            )
```
Could we add another sentence which explains that saving will bring back / recreate the file

jhaigh0 reviewed (changes requested):
Found a couple things.

First is minor, if you reject the changes made to a file, then save, then go back to notepad++ and accept the new changes, then come back to Mantid, it will say the file has been edited (I think due to np++ saving) but when you reload the file nothing has changed. Not sure if can be avoided.

Next seems more important. I closed the file in np++, deleted it in the file browser, went back to mantid and got the correct warning and clicked okay. I then saved it to recreate the file and opened it back up in np++. However, making changes in np++ now doesn't seem to affect Mantid at all.

jclarkeSTFC committed:
Add message about recreating a deleted file.

jclarkeSTFC committed:
Check file is monitored after save.

We may be saving a file that doesn't exist, so add it to the file
watcher if necessary. This can happen if the file is deleted or
renamed on disk, then saved again. When a file doesn't exist
anymore it is removed from the QFileSystemWatcher (which makes
sense).

jhaigh0 reviewed (dismissed):
This the main bug I found fixed, I think this is good to go 👍

thomashampson reviewed (changes requested):
This is a really useful addition but I think it's a bit risky to put this in just before code freeze, we probably need some time for developers to play around with it to make sure we haven't missed edge cases.

jclarkeSTFC committed:
Moved release notes to 6.9 folder

thomashampson reviewed (changes requested):
If I edit a script in workbench and save it, then click off mantid so it loses focus, then click mantid again it will tell me that the script has been modified.

jclarkeSTFC committed:
Replace monitor flag with counter

Instead of setting a bool that disabled file monitoring, use an event counter.
By using a counter we don't need to worry about when the event is handled,
we just know that X number of events are coming our way that we need to ignore.
Also fixed a small bug with Save As by only opening a tab with the new
filename if it's different to the old filename.

jclarkeSTFC committed:
Add a couple of tests

Add tests for the file watcher to check that it's detecting when files are
changed, and that it's ignoring events from when the editor itself has
saved the file.

jclarkeSTFC committed:
Change way of handling events we caused

Keep track of timestamps from saves, then we can compare the timestamp from
the notification and see if anything has changed since the save. If not, then
no user notification about the file changing. The other thing we have to
protect against is that the QFileSystemWatcher sometimes raises two events
for one file modification, so we need to check for duplicates.

jclarkeSTFC commented:
> If I edit a script in workbench and save it, then click off mantid so it loses focus, then click mantid again it will tell me that the script has been modified.

This should be fixed now, I changed the way it detects duplicate notifications, and notifications generated by itself.

jclarkeSTFC committed:
Fix test for Macs

On Macs this test will fail, seemingly because the thread with the QFileSystemWatcher
needs a little more time before it will detect external modifications to
a file. So now we do the modifications on a separate thread with a sleep.
We don't want to sleep on the QFileSystemWatcher thread because that will
also stop it doing its thing.

jclarkeSTFC committed:
Only need to check that the callback worked

No need to check the exact number of times the callback was called, because
this seems to vary depending on machine and operating system. There is
code in the widget for handling multiple calls, so we just need to make
sure that it was called at least once.

jclarkeSTFC committed:
Test that files are being watched

Test that files are added/removed from the QFileSystemWatcher correctly,
instead of checking the QFileSystemWatcher mechanism itself. The original
test that checked for the callbacks was flaky on macOS and relied on a sleep
of length that varied with the machine.

jhaigh0 reviewed (dismissed):
:shipit:

jclarkeSTFC committed:
Merge branch 'main' into 34981_detect_script_modified

thomashampson commented on buildconfig/CMake/CppCheck_Suppressions.txt.in:
```diff
@@ -1497,6 +1497,10 @@ constVariablePointer:${CMAKE_SOURCE_DIR}/qt/widgets/common/src/PropertyWidget.cp
 constVariablePointer:${CMAKE_SOURCE_DIR}/qt/widgets/common/src/PropertyWidgetFactory.cpp:43
 constVariablePointer:${CMAKE_SOURCE_DIR}/qt/widgets/common/src/PropertyWidgetFactory.cpp:44
 virtualCallInConstructor:${CMAKE_SOURCE_DIR}/qt/widgets/common/inc/MantidQtWidgets/Common/ScriptEditor.h:70
+shadowFunction:${CMAKE_SOURCE_DIR}/qt/widgets/common/src/ScriptEditor.cpp:115
+shadowFunction:${CMAKE_SOURCE_DIR}/qt/widgets/common/src/ScriptEditor.cpp:306
+shadowFunction:${CMAKE_SOURCE_DIR}/qt/widgets/common/src/ScriptEditor.cpp:506
```
```suggestion
```

cppcheck update removes these spurious defects

jclarkeSTFC committed:
Update buildconfig/CMake/CppCheck_Suppressions.txt.in

Co-authored-by: thomashampson <thomas.hampson@stfc.ac.uk>

thomashampson reviewed (changes requested):
cppcheck upgrade removes the spurious `emit` defects

thomashampson reviewed (approved):
All working well now.
